### ISSUE TITLE
To run new set (area_mean_time_series) with other diags together using API

### ISSUE THREAD

**Issue by chengzhuzhang** [E3SM-Project/e3sm_diags#241]
@dnadeau4 I'm creating this issue to better describe the problem we discussed on Tuesday. 
When resolving this issue, would you use the branch `area_mean_time_series_refine`?
The issue is that, we want to run the old sets and the new set(area_mean_time_series) together using one configuration python script in an organized way. 
On cori, I can run the `area_mean_time_series` set alone, using **`python run_api_test_ts.py -d multi_sets.cfg`**
**run_api_test_ts.py**
```
import os
from acme_diags.parameter.core_parameter import CoreParameter
from acme_diags.run import runner

param = CoreParameter()

#param.test_data_path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/time-series/CESM1-CAM5_cmip/'
param.reference_data_path = '/global/project/projectdirs/acme/acme_diags/obs_for_e3sm_diags/time-series/'
param.test_data_path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/time-series/E3SM_v1/'
param.test_name = 'e3sm_v1'
#param.seasons = ["ANN"]
param.multiprocessing = False
param.test_timeseries_input = True
param.test_start_yr = '2002'
param.test_end_yr = '2008'
param.ref_timeseries_input = True
param.ref_start_yr = '2002'
param.ref_end_yr = '2008'


#param.num_workers =  16
prefix = '/global/project/projectdirs/acme/www/zhang40/runs_with_api'
param.results_dir = os.path.join(prefix, 'area_mean_time_series_e3sm_v1_vs_obs_new_mask2')

# We're passing in this new object as well, in
# addtion to the CoreParameter object.

runner.sets_to_run = ['area_mean_time_series']
runner.run_diags([param])
```
 **multi_sets.cfg**
```
[#]
sets = ["area_mean_time_series"]
case_id = "PRECT"
variables = ["PRECT"]
ref_names = ["GPCP_v2.3"]
# reference_name = "GPCP"
regions = ["global","90S50S"]

[#]
sets = ["lat_lon"]
case_id = "GPCP_v2.3"
variables = ["PRECT"]
ref_name = "GPCP_v2.3"
reference_name = "GPCP"
seasons = ["ANN", "DJF", "MAM", "JJA", "SON"]
regions = ["global"]
test_colormap = "WhiteBlueGreenYellowRed.rgb"
reference_colormap = "WhiteBlueGreenYellowRed.rgb"
diff_colormap = "BrBG"
contour_levels = [0.5, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13, 14, 15, 16]
diff_levels = [-5, -4, -3, -2, -1, -0.5, 0.5, 1, 2, 3, 4, 5]
```
To run multiple sets, ideally we want to have configuration .py file as below
```
import os
from acme_diags.parameter.core_parameter import CoreParameter
from acme_diags.parameter.area_mean_time_series_parameter import AreaMeanTimeSeriesParameter
from acme_diags.run import runner

param = CoreParameter()

param.reference_data_path = '/global/project/projectdirs/acme/acme_diags/obs_for_e3sm_diags/climatology/'
param.test_data_path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/climatology/'
param.test_name = '20161118.beta0.FC5COSP.ne30_ne30.edison'
param.seasons = ["ANN"]
param.multiprocessing = False
prefix = '/global/project/projectdirs/acme/www/zhang40/runs_with_api'
param.results_dir = os.path.join(prefix, 'multiple_sets')


#Set specific parameters for new sets
ts_param = AreaMeanTimeSeriesParameter()
ts_param.reference_data_path = '/global/project/projectdirs/acme/acme_diags/obs_for_e3sm_diags/time-series/'
ts_param.test_data_path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/time-series/E3SM_v1/'
ts_param.test_name = 'e3sm_v1'
ts_param.test_timeseries_input = True
ts_param.test_start_yr = '2002'
ts_param.test_end_yr = '2008'
ts_param.ref_timeseries_input = True
ts_param.ref_start_yr = '2002'
ts_param.ref_end_yr = '2008'

# We're passing in this new object as well, in
# addtion to the CoreParameter object.
runner.sets_to_run = ['area_mean_time_series', 'lat_lon']
runner.run_diags([param, ts_param])
```
The goal is to have multiple sets run together use a python file like above. When we find a solution for it and have more testing for this branch. We can plan to release 2.0.0

**Comment by zshaheen** [comment]
Feel free to ping me if any of you have any questions/issues. I don't mind.

**Comment by dnadeau4** [comment]
@zshaheen Can you explain what this does?  I am trying to get parameters merged and this selections returns [] 
https://github.com/CDAT/cdp/blob/master/cdp/cdp_parser.py#L428-L459

I am trying to get this to work.
```
runner.sets_to_run = ['lat_lon', 'area_mean_time_series']
runner.run_diags([param, ts_param])
```

`cdparser.py`  final_parameters return [] and nothing is run.

Thanks!

I would like ts_param to inherit values from the CoreParameter (param).

**Comment by zshaheen** [comment]
@dnadeau4, please take a look at ["Selecting Certain Parameters"](https://e3sm-project.github.io/e3sm_diags/docs/html/defining-parameters.html#selecting-certain-parameters). Basically, this is the code that does that. In the function, the `main_parameters` is what are in `param.py` in `e3sm_diags -p param.py`. The `parameters` argument are parameters that are either:
1. The ones from `diags.cfg` in `e3sm_diags -p param.py -d diags.cfg`
2. Or if no `-d` parameter used, the default diags, which are [the `cfg` files located here](https://github.com/E3SM-Project/e3sm_diags/tree/master/acme_diags/driver/default_diags).

Let me know if you have any more questions.

**Comment by dnadeau4** [comment]
@zshaheen We want to add user-defined parameter to your API.

I was able to use `self.parser.combine_params` to merge `[param, ts_param]`.

In this example,  `parameter[0] is CoreParameter()`.  I want to add user-defined parameters to it. 

```python
        if (len(parameters) > 1):
            # Create a new parameter and copy core parameters found in parameters[0] (first in the list)
            # Then overwrite the parameters for this class
            # This will get "Global" parameters defined in parameters[0].
            for i in range(1, len(parameters)):
                param = parameters[i]
                new_param = param.__class__()  # New instance.
                self.parser.combine_params(orig_parameters=parameters[0], other_parameters=[new_param], vars_to_ignore=['seasons'])
                self.parser.combine_params(orig_parameters=param, other_parameters=[new_param], vars_to_ignore=['seasons'])
                parameters[i] = new_param
         final_params = self.get_final_parameters(parameters) 
```

I have to ignore season, due to the "select".  

In `get_final_parameters(self, parameters)` I had to remove this line
    `self._add_parent_attrs_to_children(parameters)`

otherwise the new user-defined class is being messed up by CoreParameter()

**Comment by zshaheen** [comment]
Can you explain in more detail what "We want to add user-defined parameter to your API" means? Do you want to add more parameters to `CoreParameter`, or add a new kind of parameters for a new plot set (ex: `TimeSeriesParameter`), or something else.

The idea is that you shouldn't really need to work with the stuff you're saying.

**Comment by dnadeau4** [comment]
ok here what I want to do.

https://github.com/E3SM-Project/e3sm_diags/blob/multidiags/oldset.py

**Comment by dnadeau4** [comment]
I want ts_param to overwrite some of the  CoreParameter and create an `AreaMeanTimeSeriesParameter` as well as `lat_lon`

**Comment by zshaheen** [comment]
@dnadeau4, so I'm guessing in the example that you linked, the parameters from `ts_param`, like `reference_data_path` and others are **not** being used? Like the `reference_data_path` from `param` is being used instead, even for the `area_mean_timeseries` plots?

**Comment by chengzhuzhang** [comment]
@dnadeau4 [Here](https://gist.github.com/zshaheen/bd2268732aca1a887dfa6a832bae6821) is the gist from @zshaheen, which was written and discussed before designing the new API running method. In this gist, when composing the parameter list for diags set (zonal mean 2d), we considered one scenario, this set shares most of the core parameters and have one extra set specific parameter (plev, which is defined differently than the plev in core parameters). 

@zshaheen Basically we are understanding/finalizing the logic to compose the parameters for each set. Now we are considering for area_mean_time_series set, for `reference_data_path`, `test_data_path` , etc. they are all set-specific. In current framework, my understanding is that if we define these in `area_mean_time_series_parameter.py`, and then give the parameter values in the python configuration file. These values should over-write the same parameters defined for core, when core parameters are present..is it true?

**Comment by zshaheen** [comment]
> In current framework, my understanding is that if we define these in area_mean_time_series_parameter.py, and then give the parameter values in the python configuration file. These values should over-write the same parameters defined for core, when core parameters are present..is it true?

I believe that is **incorrect**. The way that it was planned was that the `CoreParameter` has the shared parameters, while **only** the set-specific parameters were defined in the other parameter classes, like `AreaMeanTimeseriesParameter`.

So in Denis' example, since `AreaMeanTimeSeriesParameter` inherits from `CoreParameter`, the parameters such as `reference_data_path`, `test_data_path`, etc., from a `CoreParameter` object are copied over. It's [this function, `_add_parent_attrs_to_children()`](https://github.com/E3SM-Project/e3sm_diags/blob/master/acme_diags/run.py#L84) that does it.

Since you want to keep the `reference_data_path`, `test_data_path`, etc. from `AreaMeanTimeSeriesParameter`, you'll need to modify the part where the parameters from `CoreParameter` are added to those of `AreaMeanTimeSeriesParameter`, **even superceding any predefined values**. This is done in the `parameters[i] += parent` line. Note that any `Parameter` class supports addition, [seen here](https://github.com/CDAT/cdp/blob/master/cdp/cdp_parameter.py#L13).

Basically @dnadeau4, in the `_add_parent_attrs_to_children()` function, loop through all of the non-internal attributes (anything that doesn't start with `_`) of `parent` and **only** add them (via `setattr()`) if they're not present in `parameters[i]`

**Comment by chengzhuzhang** [comment]
@zshaheen Excellent! Thank you for your clarification and suggestion (I didn't know `add` is supported in `CDP`), it explains why only modifying area_mean_time_series_parameter.py didn't work. @dnadeau4 is working on and testing an approach to do superseding when there are predefined set-specific parameters.

**Comment by zshaheen** [comment]
It's basically something like this:

```python
def _add_parent_attrs_to_children(self, parameters):
    # ...
    parent = copy.deepcopy(parent)
    self._remove_attrs_with_default_values(parent)
    for attr in dir(parent):
        if not attr.startswith('_') and not hasattr(parameters[i], attr):
            # This attr of parent is a user-defined one and does not
            # already exist in the parameters[i] parameter object.
            attr_value = getattr(parent, attr)
            setattr(parameters[i], attr, attr_value)
```

**Comment by dnadeau4** [comment]
@zshaheen it worked!  Thanks for your help!  
@chengzhuzhang Can you give it a try.  my branch is `multidiags`

**Comment by chengzhuzhang** [comment]
@dnadeau4 The example configuration works. Let's test more cases for robustness. Thanks for working on it!
@zshaheen Thank you for actually coding up the approach...

**Comment by chengzhuzhang** [comment]
fixed in https://github.com/E3SM-Project/e3sm_diags/pull/248
